server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSExtractOptionsUtility.java

```diff
@@ -0,0 +1,255 @@
+    if (ctx.getClass() == CICSParser.Cics_extract_attachContext.class)
+      checkAttach((CICSParser.Cics_extract_attachContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_attributesContext.class)
+      checkAttributes((CICSParser.Cics_extract_attributesContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_certificateContext.class)
+      checkCertificate((CICSParser.Cics_extract_certificateContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_logonmessageContext.class)
+      checkLogonMsg((CICSParser.Cics_extract_logonmessageContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_processContext.class)
+      checkProcess((CICSParser.Cics_extract_processContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_tcpipContext.class)
+      checkTcpIp((CICSParser.Cics_extract_tcpipContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_tctContext.class)
+      checkTCT((CICSParser.Cics_extract_tctContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_web_clientContext.class)
+      checkWebClient((CICSParser.Cics_extract_web_clientContext) ctx);
+    else if (ctx.getClass() == CICSParser.Cics_extract_web_serverContext.class)
+      checkWebServer((CICSParser.Cics_extract_web_serverContext) ctx);
+    checkDuplicates(ctx);
+  }
+
+  private void checkAttach(CICSParser.Cics_extract_attachContext ctx) {
+    checkHasMandatoryOptions(ctx.ATTACH(), ctx, "ATTACH");
+    if (!ctx.ATTACHID().isEmpty()) {
+      checkHasIllegalOptions(ctx.CONVID(), "CONVID with ATTACHID");
+      checkHasIllegalOptions(ctx.SESSION(), "SESSION with ATTACHID");
+    } else if (!ctx.CONVID().isEmpty()) {
+      checkHasIllegalOptions(ctx.ATTACHID(), "ATTACHID with CONVID");
+      checkHasIllegalOptions(ctx.SESSION(), "SESSION with CONVID");
+    } else if (!ctx.SESSION().isEmpty()) {
+      checkHasIllegalOptions(ctx.ATTACHID(), "ATTACHID with SESSION");
+      checkHasIllegalOptions(ctx.CONVID(), "CONVID with ATTACHID");
+    }
+  }
+
+  private void checkAttributes(CICSParser.Cics_extract_attributesContext ctx) {
+    checkHasMandatoryOptions(ctx.ATTRIBUTES(), ctx, "ATTRIBUTES");
+    checkHasMandatoryOptions(ctx.STATE(), ctx, "STATE");
+
+    if (!ctx.SESSION().isEmpty()) checkHasIllegalOptions(ctx.CONVID(), "CONVID with SESSION");
+  }
+
+  private void checkCertificate(CICSParser.Cics_extract_certificateContext ctx) {
+    checkHasMandatoryOptions(ctx.CERTIFICATE(), ctx, "CERTIFICATE");
+    if (!ctx.OWNER().isEmpty()) checkHasIllegalOptions(ctx.ISSUER(), "ISSUER with OWNER");
+  }
+
+  private void checkLogonMsg(CICSParser.Cics_extract_logonmessageContext ctx) {
+    checkHasMandatoryOptions(ctx.LOGONMSG(), ctx, "LOGONMSG");
+    checkHasMandatoryOptions(ctx.LENGTH(), ctx, "LENGTH");
+    if (!ctx.INTO().isEmpty()) checkHasIllegalOptions(ctx.SET(), "SET with INTO");
+    else checkHasMandatoryOptions(ctx.SET(), ctx, "INTO or SET");
```

Comment:
`checkExactlyOneOptions("INTO, SET", ctx.INTO(), ctx.SET())` - diagnostic on fail "Exactly one of {...} must be specified"

Reply:
@chacebot Just made a branch with a function that _should_ do this in the format that Slavek laid out. Have not been able to test yet as I do not have a branch where this is needed yet. 

The branch/commit: https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/compare/development...mm-broadcom:che-che4z-lsp-for-cobol:AddMutuallyExclusiveCICSCheck

Reply:
Some of the input will be Lists of contexts...

Reply:
```
protected final <E> boolean checkMutuallyExclusiveOptions(ParserRuleContext ctx, String options, List<E>... rules) {
    boolean foundNonEmpty = false;
    for (List<E> options : rules) {
        if (options.isEmpty()) continue;
        if (foundNonEmpty) {
            throwException(..., getLocality(ctx), "Options \"" + options + "\" are mutually exclusive.");
            return false;
        }
        foundNonEmpty = true;
    }
    return true;
}
```
Similarly for the "exatctly one option" variant just generate an extra message if `foundNonEmpty == false` after the loop is done.